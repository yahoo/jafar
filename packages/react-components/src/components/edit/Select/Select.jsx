/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import SelectInternal from 'react-select';
import { isEqual } from 'lodash';
import PropTypes from 'prop-types';

export const styleOverrides = {
  control: {
    fontFamily: 'Roboto, Helvetica, Arial, sans-serif',
    borderTopWidth: '0px',
    borderLeftWidth: '0px',
    borderRightWidth: '0px',
    borderRadius: '0px',
    boxShadow: '0 0 0 0',
    backgroundColor: 'transparent',
    minHeight: '32px',
  },
  singleValue: {
    font: 'inherit',
    padding: '6px 0 7px 0',
    width: '100%',
    margin: 0,
  },
  valueContainer: {
    font: 'inherit',
    padding: '0',
  },
  clearIndicator: {
    padding: '6px 5px 5px 5px',
  },
  dropdownIndicator: {
    padding: '6px 5px 5px 5px',
  },
  menu: {
    fontFamily: 'Roboto, Helvetica, Arial, sans-serif',
  },
};

export const customStyles = {
  control: (provided) => ({
    ...provided,
    ...styleOverrides.control,
  }),
  singleValue: (provided) => ({
    ...provided,
    ...styleOverrides.singleValue,
  }),
  valueContainer: (provided) => ({
    ...provided,
    ...styleOverrides.valueContainer,
  }),
  clearIndicator: (provided) => ({
    ...provided,
    ...styleOverrides.clearIndicator,
  }),
  dropdownIndicator: (provided) => ({
    ...provided,
    ...styleOverrides.dropdownIndicator,
  }),
  menu: (provided) => ({
    ...provided,
    ...styleOverrides.menu,
  }),
};

/**
 * Represent a selection of a single item from fixed items list
 * 
 * Import <a target="_blank" 
 href="https://github.com/yahoo/jafar/blob/master/packages/react-components/src/components/edit/Select/Select.jsx">
 Select</a> from '@jafar-org/react-components/edit/Select'
 */
export default class Select extends React.Component {
  static propTypes = {
    value: PropTypes.any,
    state: PropTypes.shape({
      items: PropTypes.array,
      placeholder: PropTypes.string,
      searchable: PropTypes.bool,
      searchQuery: PropTypes.string,
    }),
    disabled: PropTypes.bool,
    required: PropTypes.bool,
    onValueChange: PropTypes.func.isRequired,
    onStateChange: PropTypes.func,
  };

  static defaultProps = {
    value: undefined,
    state: {
      items: [],
      placeholder: 'Search',
      searchable: false,
      searchQuery: '',
    },
    disabled: false,
    required: false,
  };

  render() {
    // null - fixes underline controlled VS uncontrolled issue when value turns to undefined
    let selected = (this.props.state.items || []).find(x => isEqual(x.value, this.props.value)) || null;

    return (
      <SelectInternal
        value={selected}
        onChange={this.onValueChange}
        options={this.props.state.items}
        placeholder={this.props.state.placeholder}
        isDisabled={this.props.disabled}
        isClearable={!this.props.required}
        isSearchable={!!this.props.state.searchable}
        inputValue={this.props.state.searchQuery}
        onInputChange={this.onSearchChange}
        styles={customStyles}
      />
    );
  }

  onValueChange = (selected) => {
    let newValue = selected ? selected.value : selected;
    this.props.onValueChange(newValue);
  }

  onSearchChange = (query) => {
    const newState = Object.assign({}, this.props.state);
    newState.searchQuery = query;
    this.props.onStateChange(newState);
  }
}
