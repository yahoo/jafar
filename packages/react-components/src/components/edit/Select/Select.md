<h3>Examples</h3>

<h4>Usage in jafar form</h4>

```javascript
const Form = require('@jafar/react-form/Form').default;
const Field = require('@jafar/react-form/Field').default;

const model = {
  id: 'simple',
  fields: {
    bestFriend: {
      label: 'Best Friend',
      path: 'bestFriend',
      component: {
        name: 'Select',
        state: {
        items: [{
                value: '1',
                label: 'Ross Geller'
            }, {
                value: '2',
                label: 'Monica Geller'
            }, {
                value: '3',
                label: 'Rachel Green'
            }, {
                value: '4',
                label: 'Chandler Bing'
            }, {
                value: '5',
                label: 'Joey Tribbiani'
            }, {
                value: '6',
                label: 'Phoebe Buffay'
            }],
        },
      },
      required: true,
      disableTerm: {
          name: 'disableMe',
      },
      excludeTerm: {
          name: 'excludeMe',
      },
    },
  },
  data: {
      bestFriend: '2', // remove to see initial value undefined, and required error
  },
};

const resources = {
  components: { 
    Select: {
      renderer: Select,
    }, 
  },
  terms: {
      disableMe: { func: () => false }, // convert to true to see the field disabled
      excludeMe: { func: () => false }, // convert to true to see the field excluded
  },
};

<Form model={model} resources={resources}>
    <Field id='bestFriend' />
</Form>
```

<h4>Simple</h4>

```javascript
initialState = { 
  value: '3',
  state: {
    items:[{
        value: '1',
        label: 'Ross Geller'
    }, {
        value: '2',
        label: 'Monica Geller'
    }, {
        value: '3',
        label: 'Rachel Green'
    }, {
        value: '4',
        label: 'Chandler Bing'
    }, {
        value: '5',
        label: 'Joey Tribbiani'
    }, {
        value: '6',
        label: 'Phoebe Buffay'
    }],
  },
};

<Select
  value={state.value}
  state={state.state}
  onValueChange={(value) => {
    setState({ value });
  }}
/>
  ```

<h4>Value as object</h4>

```javascript
initialState = { 
  value: { id: '3', first: 'Rachel', last: 'Green' },
  state: {
    items:[{
      value: { id: '1', first: 'Ross', last: 'Geller' },
      label: 'Ross Geller'
    }, {
      value: { id: '2', first: 'Monica', last: 'Geller' },
      label: 'Monica Geller'
    }, {
      value: { id: '3', first: 'Rachel', last: 'Green' },
      label: 'Rachel Green'
    }, {
      value: { id: '4', first: 'Chandler', last: 'Bing' },
      label: 'Chandler Bing'
    }, {
      value: { id: '5', first: 'Joey', last: 'Tribbiani' },
      label: 'Joey Tribbiani'
    }, {
      value: { id: '6', first: 'Phoebe', last: 'Buffay' },
      label: 'Phoebe Buffay'
    }],
  },
};

<Select
  value={state.value}
  state={state.state}
  onValueChange={(value) => {
    setState({ value });
  }}
/>
```


<h4>Search</h4>

```javascript
initialState = { 
  value: '3',
  state: {
    searchable: true,
    items:[{
        value: '1',
        label: 'Ross Geller'
    }, {
        value: '2',
        label: 'Monica Geller'
    }, {
        value: '3',
        label: 'Rachel Green'
    }, {
        value: '4',
        label: 'Chandler Bing'
    }, {
        value: '5',
        label: 'Joey Tribbiani'
    }, {
        value: '6',
        label: 'Phoebe Buffay'
    }],
  },
};

<Select
  value={state.value}
  state={state.state}
  onValueChange={(value) => {
    setState({ value });
  }}
/>
```

<h4>Required (cant unselect)</h4>

```javascript
initialState = { 
  value: '3',
  state: {
    items:[{
      value: '1',
      label: 'Ross Geller'
    }, {
      value: '2',
      label: 'Monica Geller'
    }, {
      value: '3',
      label: 'Rachel Green'
    }, {
      value: '4',
      label: 'Chandler Bing'
    }, {
      value: '5',
      label: 'Joey Tribbiani'
    }, {
      value: '6',
      label: 'Phoebe Buffay'
    }],
  },
};

<Select
  value={state.value}
  state={state.state}
  required={true}
  onValueChange={(value) => {
    setState({ value });
  }}
/>
  ```
<h4>Disabled</h4>

```javascript
initialState = { 
  value: '3',
  state: {
    items:[{
        value: '1',
        label: 'Ross Geller'
    }, {
        value: '2',
        label: 'Monica Geller'
    }, {
        value: '3',
        label: 'Rachel Green'
    }, {
        value: '4',
        label: 'Chandler Bing'
    }, {
        value: '5',
        label: 'Joey Tribbiani'
    }, {
        value: '6',
        label: 'Phoebe Buffay'
    }],
  },
};

<Select
  value={state.value}
  state={state.state}
  disabled={true}
  onValueChange={(value) => {
    setState({ value });
  }}
/>
```

<h4>Value undefined</h4>

```javascript
initialState = { 
  value: undefined,
  state: {
    items:[{
        value: '1',
        label: 'Ross Geller'
    }, {
        value: '2',
        label: 'Monica Geller'
    }, {
        value: '3',
        label: 'Rachel Green'
    }, {
        value: '4',
        label: 'Chandler Bing'
    }, {
        value: '5',
        label: 'Joey Tribbiani'
    }, {
        value: '6',
        label: 'Phoebe Buffay'
    }],
  },
};

<Select
  value={state.value}
  state={state.state}
  onValueChange={(value) => {
    setState({ value });
  }}
/>
```