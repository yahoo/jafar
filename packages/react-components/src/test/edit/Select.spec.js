/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import { shallow } from 'enzyme';
import Select from '../../components/edit/Select';
import { mapper } from '../../components/edit/Select/Select';
  
describe('Select', () => {
  let componentProps;
  
  const jafarProps = {
    value: 'ocean',
    disabled: false,
    required: false,
    state: {
      searchable: true,
      searchQuery: 'oc',
      placeholder: 'Color...',
      items:[
        { value: 'ocean', label: 'Ocean' },
        { value: 'blue', label: 'Blue' },
        { value: 'purple', label: 'Purple' },
      ],
    },
    onValueChange: jest.fn(),
    onStateChange: jest.fn(),
  };

  const expectedProps = {
    value: { value: 'ocean', label: 'Ocean' },
    options: jafarProps.state.items,
    placeholder: 'Color...',
    isDisabled: false,
    isClearable: true,
    isSearchable: true,
    inputValue: 'oc',
    styles: expect.any(Object),
    onChange: expect.any(Function),
    onInputChange: expect.any(Function),
  };
   
  beforeEach(() => {
    componentProps = mapper(jafarProps);
  });
  
  describe('mapper', () => {
    it('return correct props', () => {
      expect(componentProps).toEqual(expectedProps);
    });
    
    it('onChange - call onValueChange with correct value', () => {
      const selected = expectedProps.options[1];
      componentProps.onChange(selected);
      expect(jafarProps.onValueChange).toHaveBeenCalledWith(jafarProps.state.items[1].value);
    });
 
    it('onChange - call onValueChange with undefined value', () => {
      componentProps.onChange(undefined);
      expect(jafarProps.onValueChange).toHaveBeenCalledWith(undefined);
    });
 
    it('onInputChange - call onStateChange with correct value', () => {
      componentProps.onInputChange('ra');
      expect(jafarProps.onStateChange).toHaveBeenCalledWith(expect.any(Function));
      const updater = jafarProps.onStateChange.mock.calls[0][0];
      expect(updater({ state: { items: [] } })).toEqual({ items: [], searchQuery: 'ra' });
    });
  });
  
  describe('component', () => {
    it('renders ok', () => {
      const component = shallow(<Select {...jafarProps} />);
      expect(component.props()).toMatchObject(expectedProps);
    });
  });
});
 
