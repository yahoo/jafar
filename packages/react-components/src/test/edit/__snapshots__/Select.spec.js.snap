// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> Should render provided data - disabled 1`] = `
<StateManager
  defaultInputValue=""
  defaultMenuIsOpen={false}
  defaultValue={null}
  isClearable={true}
  isDisabled={true}
  isSearchable={false}
  onChange={[Function]}
  onInputChange={[Function]}
  styles={
    Object {
      "clearIndicator": [Function],
      "control": [Function],
      "dropdownIndicator": [Function],
      "menu": [Function],
      "menuList": [Function],
      "singleValue": [Function],
      "valueContainer": [Function],
    }
  }
  value={null}
/>
`;

exports[`<Select /> Should render provided data - required 1`] = `
<StateManager
  defaultInputValue=""
  defaultMenuIsOpen={false}
  defaultValue={null}
  isClearable={false}
  isDisabled={false}
  isSearchable={false}
  onChange={[Function]}
  onInputChange={[Function]}
  styles={
    Object {
      "clearIndicator": [Function],
      "control": [Function],
      "dropdownIndicator": [Function],
      "menu": [Function],
      "menuList": [Function],
      "singleValue": [Function],
      "valueContainer": [Function],
    }
  }
  value={null}
/>
`;

exports[`<Select /> Should render provided data 1`] = `
<StateManager
  defaultInputValue=""
  defaultMenuIsOpen={false}
  defaultValue={null}
  isClearable={true}
  isDisabled={false}
  isSearchable={false}
  onChange={[Function]}
  onInputChange={[Function]}
  styles={
    Object {
      "clearIndicator": [Function],
      "control": [Function],
      "dropdownIndicator": [Function],
      "menu": [Function],
      "menuList": [Function],
      "singleValue": [Function],
      "valueContainer": [Function],
    }
  }
  value={null}
/>
`;
